### ISSUE TITLE
Python 3.12 segfaults when called from different OpenMP threads in a C extension

### ISSUE THREAD

**Issue by eugvas** [python/cpython#111034]
# Crash report

### What happened?

I came across a problem when a C extension module calls any Py*** function from multiple OpenMP threads. The problem only exists on Python 3.12 installed from Anaconda (so perhaps it's their problem, not Python itself?), both on Linux or Mac OS.

Here is a minimal (?) example: first the C extension module, then the Python stub for compiling it.
`testomp.c`
```C
#include "Python.h"
#include "omp.h"
PyObject* test(PyObject* self, PyObject* obj) {
    int count = 10, i, k;
    printf("spawning %i threads\n", omp_get_max_threads());
#pragma omp parallel for schedule(dynamic,1)
    for(i=0; i<count; i++) {
        for(k=0; k<i*1234%5*67890; k++) {}  // add some delay
#pragma omp critical
        {
            printf("enter critical section for iter %i, thread %i\n", i, omp_get_thread_num());
            PyObject* str = PyObject_Str(obj);
            if(str) {
                printf("leave critical section for iter %i, thread %i: %s\n", i, omp_get_thread_num(),
                    PyUnicode_AsUTF8(str));
                Py_DECREF(str);
            }
        }
    }
    Py_RETURN_NONE;
}

PyMODINIT_FUNC PyInit_testomp(void) {
    static PyMethodDef methods[] = { {"test", test, METH_O, NULL}, {NULL, NULL, 0, NULL} };
    static PyModuleDef moduledef = { PyModuleDef_HEAD_INIT, "testomp", "", -1, methods };
    return PyModule_Create(&moduledef);
}
```
`setup.py` for Linux
```python
from setuptools import setup, Extension
setup(script_args=['build_ext'], ext_modules=[Extension('testomp', sources=['testomp.c'],
    extra_compile_args=['-fopenmp'],
    extra_link_args=['-fopenmp'])])
```
`setup.py` for MacOS compiled with clang (hence some wiggles to make it understand OpenMP directives) and linked against `libomp.dylib` provided by Homebrew
```python
from setuptools import setup, Extension
setup(script_args=['build_ext'], ext_modules=[Extension('testomp', sources=['testomp.c'],
    extra_compile_args=['-I/usr/local/opt/libomp/include', '-Xpreprocessor', '-fopenmp'],
    extra_link_args=['-L/usr/local/opt/libomp/lib', '-lomp'])])
```

If I compile the extension module `testomp`, then run something like
`python -c "import testomp; testomp.test('bla')"`
the result looks something like this in Python 3.11 (and presumably any previous version)
```
spawning 16 threads
enter critical section for iter 0, thread 0
leave critical section for iter 0, thread 0: bla
enter critical section for iter 4, thread 2
leave critical section for iter 4, thread 2: bla
enter critical section for iter 5, thread 4
leave critical section for iter 5, thread 4: bla
enter critical section for iter 1, thread 3
leave critical section for iter 1, thread 3: bla
enter critical section for iter 3, thread 8
leave critical section for iter 3, thread 8: bla
enter critical section for iter 6, thread 15
leave critical section for iter 6, thread 15: bla
enter critical section for iter 2, thread 12
leave critical section for iter 2, thread 12: bla
enter critical section for iter 7, thread 14
leave critical section for iter 7, thread 14: bla
enter critical section for iter 8, thread 11
leave critical section for iter 8, thread 11: bla
enter critical section for iter 9, thread 13
leave critical section for iter 9, thread 13: bla
```
But when run with Python 3.12, it segfaults as soon as the thread index differs from zero (here example from Mac, same happens on Linux):
```
spawning 16 threads
enter critical section for iter 0, thread 0
leave critical section for iter 0, thread 0: bla
enter critical section for iter 1, thread 0
leave critical section for iter 1, thread 0: bla
enter critical section for iter 2, thread 0
leave critical section for iter 2, thread 0: bla
enter critical section for iter 3, thread 0
leave critical section for iter 3, thread 0: bla
enter critical section for iter 4, thread 0
leave critical section for iter 4, thread 0: bla
enter critical section for iter 5, thread 12
zsh: segmentation fault  python -c "import testomp; testomp.test('bla')"
```

As one could see, access to Python functions is protected by a critical section (`#pragma omp critical`), so it's not caused by calling these functions from multiple threads simutaneously (also confirmed by the ordered appearance of "enter/leave critical section" messages in the output for Py 3.11 and earlier). It seems that the crash is caused specifically by accessing Python from a non-main thread.
Am I doing something wrong here, how should such access be protected? or is this a bug introduced in the latest version?

### CPython versions tested on:

3.12

### Operating systems tested on:

Linux, macOS

**Comment by ronaldoussoren** [comment]
AFAIK this is a bug in your code, the ``amp parallel`` loop runs the body of the loop in multiple threads. Because of that you should acquire and release the GIL in the body of the loop.

**Comment by eugvas** [comment]
thanks for a hint! 
Following the recommendation from https://docs.python.org/3/c-api/init.html#non-python-created-threads, I tried to add
`PyGILState_STATE state = PyGILState_Ensure();`
before calling the Python functions (at the beginning of the critical section), and 
`PyGILState_Release(state);`
at the end of this section (after doing all business with Python), but this leads to a deadlock (both in Python 3.11 and 3.12), regardless of whether I additionally protect this section by a `#pragma omp critical` directive or not.

I assume that an OpenMP critical section is sufficient to ensure that no two threads can call Python routines at the same time, and this seems to work fine on previous version of Python. I am puzzled why adding the Ensure/Release calls leads to a deadlock, or is this not the way it should be done?

**Comment by eugvas** [comment]
Many thanks, this now works properly (and even does not need `#pragma omp critical`, as the `PyGILState_Ensure()` call itself creates a barrier). 
I confess that the documentation on this issue is fairly confusing and does not indicate that these two pairs of functions have to be used in this way: I tried putting Py_BEGIN/END_ALLOW_THREADS where they are in the above code _without_ corresponding calls to PyGILState_Ensure/Release, and it hangs; tried it the other way round, and it segfaults.. it didn't occur to me that one needs to put them exactly in this sequence! 
I found one page https://www.geeksforgeeks.org/releasing-gil-and-mixing-threads-from-c-and-python/ where this example is shown, but even there it is not stressed that it is crucial to do both (and in many SO threads this aspect is missed).
The only remaining question for me is why this all worked without hassle in previous versions (<=3.11).
thanks again and I think this issue may be closed (though I would certainly welcome any follow-up action on improving the docs, and would be happy to hear any comments on my remaining question).
